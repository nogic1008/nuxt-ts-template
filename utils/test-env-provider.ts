/* eslint-disable no-process-env */
/** Provides process.env for Testing. */
export default class TestEnvProvider<T> {
  /** Stored process.env */
  private readonly storedEnv: NodeJS.ProcessEnv
  private readonly keys: Set<Extract<keyof T, string>>

  constructor(...keys: Extract<keyof T, string>[]) {
    this.storedEnv = { ...process.env }
    this.keys = new Set(keys)
  }

  initialize(obj?: Partial<T>) {
    const env = { ...this.storedEnv }
    this.keys.forEach((k) => delete env[k])
    process.env = { ...env, ...obj }
  }

  restore() {
    process.env = { ...this.storedEnv }
  }
}
