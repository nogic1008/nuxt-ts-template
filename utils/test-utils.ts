/** Generate [0-9a-z]{length} string. */
export const generateRandomString = (length: number) =>
  [...Array(Number.isInteger(length) ? length : 12)]
    .map(() => (~~(Math.random() * 36)).toString(36))
    .join('')
