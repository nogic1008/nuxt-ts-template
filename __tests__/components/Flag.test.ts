import { createLocalVue, mount, shallowMount } from '@vue/test-utils'

import FlagComponent from '~/components/Flag.vue'

import { randomString as random } from '../utils'

const localVue = createLocalVue()

type FlagProps = {
  iso: string
  title?: string
  squared?: boolean
}

describe('components/Flag.vue', () => {
  let wrapper: ReturnType<typeof mount>
  const propsData: FlagProps = { iso: 'ja' }
  beforeEach(() => {
    wrapper = shallowMount(FlagComponent, { localVue, propsData })
  })

  test('is a Vue instance', () => {
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  test('renders correctly', () => {
    const wrapper = mount(FlagComponent, {
      localVue,
      propsData: { ...propsData, title: 'Japan', squared: true }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  test('sets { squared: false, title: undefined } default', () => {
    expect(wrapper.props()).toStrictEqual({
      ...propsData,
      squared: false,
      title: undefined
    })
  })

  test.each([
    ['flag-icon-ja', 'ja'],
    ['flag-icon-us', 'US'],
    ['flag-icon-foo', 'Foo']
  ])('has %s class if iso is %s', async (expectedClass, iso) => {
    wrapper.setProps({ iso })
    await wrapper.vm.$nextTick()
    expect(wrapper.find('span').classes()).toContain(expectedClass)
  })

  test('has flag-icon-squared class if squared is true', async () => {
    wrapper.setProps({ squared: true })
    await wrapper.vm.$nextTick()
    expect(wrapper.find('span').classes()).toContain('flag-icon-squared')
  })

  test('has not flag-icon-squared class if squared is false', async () => {
    wrapper.setProps({ squared: false })
    await wrapper.vm.$nextTick()
    expect(wrapper.find('span').classes()).not.toContain('flag-icon-squared')
  })

  describe('title attribute', () => {
    test('equals iso prop if title is undefined', async () => {
      const randomString = random(10)
      wrapper.setProps({ iso: randomString })
      await wrapper.vm.$nextTick()
      expect(wrapper.find('span').attributes().title).toBe(randomString)
    })
    test('equals title prop', async () => {
      const randomString = random(10)
      wrapper.setProps({ title: randomString })
      await wrapper.vm.$nextTick()
      expect(wrapper.find('span').attributes().title).toBe(randomString)
    })
  })
})
