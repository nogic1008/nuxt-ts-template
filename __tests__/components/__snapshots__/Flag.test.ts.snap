// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Flag.vue renders correctly 1`] = `
<span
  class="flag-icon flag-icon-squared flag-icon-ja"
  title="Japan"
/>
`;
